File: services/nuldoc/nuldoc-src/components/StaticScript.ts
import { join } from "@std/path";
import { Config } from "../config.ts";
import { Element, script } from "../dom.ts";
import { calculateFileHash } from "./utils.ts";

export default async function StaticScript(
  { site, fileName, type, defer, config }: {
    site?: string;
    fileName: string;
    type?: string;
    defer?: "true";
    config: Config;
  },
): Promise<Element> {
  const filePath = join(
    Deno.cwd(),
    config.locations.staticDir,
    site || "_all",
    fileName,
  );
  const hash = await calculateFileHash(filePath);
  const attrs: Record<string, string> = { src: `${fileName}?h=${hash}` };
  if (type) attrs.type = type;
  if (defer) attrs.defer = defer;
  return script(attrs);
}

File: services/nuldoc/lib/nuldoc/components/static_script.rb
module Nuldoc
  module Components
    class StaticScript < DOM::HTMLBuilder
      def initialize(file_name:, config:, site: nil, type: nil, defer: nil)
        super()
        @file_name = file_name
        @config = config
        @site = site
        @type = type
        @defer = defer
      end

      def build
        file_path = File.join(Dir.pwd, @config.locations.static_dir, @site || '_all', @file_name)
        hash = ComponentUtils.calculate_file_hash(file_path)
        attrs = { src: "#{@file_name}?h=#{hash}" }
        attrs[:type] = @type if @type
        attrs[:defer] = @defer if @defer
        script(**attrs)
      end
    end
  end
end
